### PR TITLE
Added fullDay default duration for events

### PR DESCRIPTION
**Pull request by Jacob-Daniel**
**Changes**
Implemented fullDay functionality for the Calendar to support events spanning multiple days.
Updated the default behaviour to handle multi-day events in a 'Booking' collection type.

**Testing**
Tested functionality in the month view successfully.
Bookings were correctly displayed across the appropriate duration of days.
Clicking on an event correctly redirects to the corresponding Booking details page in Strapi.

**Error**
Encountered an error when switching between calendar views (e.g., from month to week or day view), unrelated to the fullDay changes. The error (Uncaught TypeError: Cannot read properties of undefined (reading 'end')) was reproduced on a fresh install.
A separate issue will be opened to track this bug.

**Versions**
Strapi version: 4.25.11
Calendar plugin version: ^0.1.2
Node version: v20.17.0

### CODE REVIEW

File: admin/src/pages/settings.js [hunk 1]
```diff
@@ -298,6 +298,12 @@ function Settings() {
                         defaultMessage: '2 Hours',
                       })}
                     </Option>
+					<Option value={1440}>
```
Comment (LuisRodriguezLD) [inline admin/src/pages/settings.js:301]: can you fix the indentation?

File: server/services/service.js [hunk 2]
```diff
@@ -81,7 +81,9 @@ module.exports = () => ({
       startDate: x[config.startField],
       endDate: config.endField
         ? x[config.endField]
-        : moment(x[config.startField]).add(config.defaultDuration, 'minutes'),
+        : config.defaultDuration === 'fullDay'
```
Comment (LuisRodriguezLD) [inline server/services/service.js:84]: default duration will never be fullDay, you would need to compare to that value which is `1440`

File: server/services/service.js [hunk 3]
```diff
@@ -81,7 +81,9 @@ module.exports = () => ({
       startDate: x[config.startField],
```
Comment (LuisRodriguezLD) [inline server/services/service.js:81]: if the event is fullDay, shouldn't it also affect the start?

**Review by LuisRodriguezLD — CHANGES_REQUESTED** [review]
thanks for the PR, 
I've left a few comments, let me know if you have questions

File: admin/src/pages/settings.js [hunk 4]
```diff
@@ -61,6 +61,7 @@ function Settings() {
     titleField: null,
     colorField: null,
     defaultDuration: 30,
+    defaultStartTime: '12:00',
```
Comment (LuisRodriguezLD) [inline admin/src/pages/settings.js:64]: as of now, there's no way for users to edit this so let's remove it and uses moment instead

**Review by LuisRodriguezLD — CHANGES_REQUESTED** [review]

File: server/services/service.js [hunk 5]
```diff
@@ -74,16 +74,21 @@ module.exports = () => ({
...
+      const isMultiDay = endDate && moment(endDate).isAfter(startDate); 
+      const startDateTime = `${startDate}T00:00:00`; 
+      const endDateTime = isMultiDay ? `${endDate}T00:00:00` : moment(startDateTime).add(config.defaultDuration, 'minutes').format(); 
+      return {
+         id: x.id,
+         title: config.titleField ? x[config.titleField] : x[config.startField],
+         startDate: startDateTime,
+         endDate: endDateTime,
+         color: config.colorField ? x[config.colorField] : null,
+         allDay: isMultiDay,
```
[the remark this answers is not in the document]
Comment (Jacob-Daniel) [reply server/services/service.js:89]: I’m a bit confused by the request to remove the allDay attribute. The primary aim of this pull request was to ensure that full-day and multi-day events are accurately represented in the calendar, and this relies on the allDay attribute. Without it, handling multi-day events, particularly across time zones, becomes more challenging.

Since the allDay attribute will be required in the upcoming version, wouldn’t it make sense to retain it in this PR for future compatibility and to ensure proper rendering now? I’d appreciate it if you could clarify the reasoning behind its removal, given its role in managing these event types.

Thanks

[the remark this answers is not in the document]
Comment (LuisRodriguezLD) [reply server/services/service.js:89]: Apologies.
Upon further inspection it turns out the allDay attribute **is** needed (Demo [here](https://codesandbox.io/p/sandbox/lvhdrt?file=%2Fdemo-data%2Fappointments.js))

File: admin/src/pages/settings.js
[the remark this answers is not in the document]
Comment (LuisRodriguezLD) [reply admin/src/pages/settings.js:1]: Indentation is perfect. Thank you!

File: server/services/service.js [hunk 6]
```diff
@@ -74,16 +74,23 @@ module.exports = () => ({
...
-    return dataFiltered.map((x) => ({
-      id: x.id,
-      title: config.titleField ? x[config.titleField] : config.startField,
-      startDate: x[config.startField],
-      endDate: config.endField
-        ? x[config.endField]
-        : moment(x[config.startField]).add(config.defaultDuration, 'minutes'),
-      color: config.colorField ? x[config.colorField] : null,
-    }));
+    return dataFiltered.map((x) => {
```
Comment (LuisRodriguezLD) [inline server/services/service.js:77]: I think the way variables names are defined is a bit confusing.

I feel startDate & endDate should be as before considering all prev rules,
this should handle cases where the defaultDuration is 1440 minutes
```js
let startDate = x[config.startField];
let endDate = config.endField
         ? x[config.endField]
         : moment(x[config.startField]).add(config.defaultDuration, 'minutes');
```

with that out of the way, we can now check if it's an allDay event using [moment.diff](https://momentjs.com/docs/#/displaying/difference/) and adjust the variables if it's a 24h event
```js
const allDay = moment(endDate).diff(startDate, "hours") === 24;
if (addDay) {
 startDate = moment(startDate).startOf('day');
 endDate = moment(startDate).endOf('day');
}
```

this should handle cases where 
1. the default duration is 24 hours 
2. where the endDate is 24 hours after the start date

I guess my question would be, do we want to span across multiple days?
perhaps an even better solution is to have an `allDay` boolean as part of the content (this is how regular calendars work)

**Review by LuisRodriguezLD — CHANGES_REQUESTED** [review]

File: server/services/service.js (hunk 6 above)
Comment (Jacob-Daniel) [reply server/services/service.js:77]: Hi Luis,

Yes, renaming `isMultiDay` back to `allDay` aligns better with FullCalendar’s convention and expected behaviour. I think I was incorrectly too focused on the `allDay` field determining the desired result of my use case.. 

The `allDay` flag determines whether an event spans the entire day, while the startDate and endDate manage how long the event spans, including multi-day events, so that is easily managed in our logic.

So, for multi-day events, it's primarily the endDate that defines if an event spans multiple days, as when viewing in month view. The `allDay` flag can be used for full-day events (whether single or multi-day). If `allDay` is set to null, FullCalendar will rely on the default values, and if the `endDate` extends beyond the start date, the calendar will visually span across those days.

Multiple Days - this is definitely required for my use case, where I am designing a campsite booking application! 

I did not test your code, but I think this address your concerns while achieving my desired use case functionality:
   
```
   return dataFiltered.map((x) => {
      let startDate = x[config.startField];
      let endDate = config.endField
        ? x[config.endField]
        : moment(x[config.startField]).add(config.defaultDuration, 'minutes').format();

      // Check if the default duration equals 1440 minutes, meaning it's a full-day event
      const fullDay = config.defaultDuration === 1440;

      if (fullDay) {
        startDate = moment(startDate).startOf('day').format();
        endDate = moment(endDate).endOf('day').format();
      }

      return {
        id: x.id,
        title: config.titleField ? x[config.titleField] : x[config.startField],
        startDate: startDate,
        endDate: endDate,
        color: config.colorField ? x[config.colorField] : null,
        allDay: fullDay, // Full-day flag
      };
    });
```
Please see screen shots of working example.
strapi-fullcalendar-allday-multiple-days-day-view
strapi-fullcalendar-allday-span-multiple-days-month-view
strapi-fullcalendar-allday-span-multiple-days-week-view

Comment (LuisRodriguezLD) [reply server/services/service.js:77]: Thank  you Jacob, 
I agree with everything.

Please push your changes when you can :)

**Comment by LuisRodriguezLD** [comment]
Code looks good but the build failed, could you run `npm run lint:fix` and push the changes?

File: pnpm-lock.yaml
Comment (LuisRodriguezLD) [inline pnpm-lock.yaml:1]: This field should not be here. We use yarn.
Please remove

File: package.json
Comment (LuisRodriguezLD) [inline package.json:1]: I think the lint script updated this file by mistake. could you revert the change?

File: package-lock.json
Comment (LuisRodriguezLD) [inline package-lock.json:1]: This file also shouldn't exist because we use yarn

**Review by LuisRodriguezLD — CHANGES_REQUESTED** [review]
Code is solid. Let's just clean up before we merge. 
Thank you for your time

File: pnpm-lock.yaml
Comment (Jacob-Daniel) [reply pnpm-lock.yaml:1]: Done, sorry for that ..

File: package.json
Comment (Jacob-Daniel) [reply package.json:1]: I think my last push looks okay now, again sorry for that ..

**Comment by Jacob-Daniel** [comment]
Louis, I think you mentioned a new release? Do you have an out-line of goals/ objectives? Just, as you know I am using your plugin and there maybe features I would like to contribute. Thanks for your recent help.

**Comment by LuisRodriguezLD** [comment]
Code is flawless, thank you Jacob.
Regarding the next major version, I published a beta a few weeks ago, the work is [here](https://github.com/offset-dev/strapi-calendar/tree/v1). 
I don't have specific goals or objectives other than 
1. change calendar library
2. TS support everywhere
3. use new strapi plugin sdk
4. strapi5 compat

Fee free to join the conversation [here](https://github.com/offset-dev/strapi-calendar/issues/42)

**Comment by Jacob-Daniel** [comment]
Hi Luis,

I’m looking to test the version 1 beta of the FullCalendar Strapi plugin, and I noticed that the fullDay duration option is not available in this version. I wanted to clarify a few things:

Strapi V4 vs V5 Focus: Is the develop branch primarily focused on supporting Strapi v4, while version 1 targets Strapi v5, or is there a goal to eventually consolidate the two?

Calendar Options: Given that there are relatively few calendar options available in the plugin, is the intention for developers to contribute additional FullCalendar library options (such as validRange, initialDate, etc.)?

Specifically, my current need is to control the calendar range like so:

```
validRange={{
  start: '2024-06-01',
  end: '2024-08-31',
}}
initialDate="2024-06-01"
```

If you prefer, I can post this in the version 1 thread, but I wanted to clarify these points with you first.

Thanks for your help!

**Comment by LuisRodriguezLD** [comment]
Hey Jacob, thanks for reaching out!

> Strapi V4 vs V5 Focus: Is the develop branch primarily focused on supporting Strapi v4, while version 1 targets Strapi v5, or is there a goal to eventually consolidate the two?
Think of the `v1` branch as a feature branch to make the plugin Strapi5 ready.
I haven't heard issues so I'm creating a PR towards main & develop later this week

> Calendar Options: Given that there are relatively few calendar options available in the plugin, is the intention for developers to contribute additional FullCalendar library options (such as validRange, initialDate, etc.)?
That is correct. As of know, most features from v0 are available in v1 (except for the full day feature).
Developers should contribute using the FullCalendar library

You can create a new feature from the v1 branch and as soon as it lands in develop you can create a PR towards develop

**Comment by Jacob-Daniel** [comment]
Thanks Luis, that makes sense. I will comment on the v1 thread stating intentions.
